adapter/service/virtualmeter/virtual_meter_manager.go

```diff
@@ -0,0 +1,332 @@
+package virtualmeter
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/futurehomeno/cliffhanger/adapter"
+	"github.com/futurehomeno/cliffhanger/task"
+)
+
+type (
+	Manager interface {
+		Add(addr string, modes map[string]float64, unit string) error
+		Remove(addr string) error
+		Modes(addr string) (map[string]float64, error)
+		SetReportingInterval(duration time.Duration) error
+		ReportingInterval() time.Duration
+
+		// RegisterDevice saves service by address that will be added to the thing on respective fimp message or
+		// add service to the thing immediately if it was initialised previously.
+		RegisterDevice(thing adapter.Thing, addr string, service adapter.Service) error
+		// Update updates a virtual meter for a device by a given addr with a new mode and level.
+		Update(addr, mode string, level float64) error
+		// Report returns a value to report based on a provided unit.
+		Report(addr, unit string) (float64, error)
+	}
+
+	virtualMeterManager struct {
+		lock             sync.RWMutex
+		ad               adapter.Adapter
+		serviceTemplates map[string]adapter.Service
+		storage          Storage
+		taskManager      task.Manager
+	}
+)
+
+var _ Manager = &virtualMeterManager{}
+
+// NewVirtualMeterManager creates a new virtual meter manager with basic initialisation.
+func NewVirtualMeterManager(workdir string) Manager {
+	return &virtualMeterManager{
+		lock:             sync.RWMutex{},
+		serviceTemplates: make(map[string]adapter.Service),
+		storage:          NewStorage(workdir),
+	}
+}
+
+// WithAdapter adds a provided adapter to the provided virtual meter manager. Used to avoid circular dependencies.
+func WithAdapter(meter Manager, ad adapter.Adapter) Manager {
```

Comment:
Don't understand why it's like this? Why this is not a public method of the Manager interface?